04 - Poker.playground/Contents.swift

```diff
@@ -6,18 +6,286 @@
+            return .quadruple
+        }
+            
+        else if (HasPair(cards).hasSecondPair && HasPair(cards).hasTriple && (HasPair(cards).highPairRank != HasPair(cards).highTripleRank || HasPair(cards).secondPairRank != HasPair(cards).highTripleRank )) {
+            return .fullHouse
+        }
+        else if HasStraight(cards) {
+            return .straight
+        }
+        
+        else if HasFlush(cards).hasFlush {
+            return .flush
+            }
+            
+        else if HasPair(cards).hasTriple {
+            return .triplet
+        }
+            
+        else if HasPair(cards).hasSecondPair {
+            return .twoPair
+        }
+        
+        else if HasPair(cards).hasPair {
+            return .pair
+        }
+        
+    
+        else {
+           return .highcard
+        }
+        
+    }
+    
+    
+}
 
 
 
 
+func HasFlush (cards : [Card]) -> (hasFlush: Bool, highestRank: Rank) {
+    var hasFlush: Bool = false
+    let reference = cards[0].suit
+    let highestCard = cards[0].rank
+    
+    for i in 1...4 {
+        
+        
+        if cards[i].suit != reference {
+        
+            break
+            
+        }
+        if i==4 { hasFlush = true }
+    
+    }
+    
+    return (hasFlush, highestCard)
+}
+
+func HasPair (cards : [Card]) -> (hasPair: Bool, hasSecondPair: Bool, hasTriple: Bool, hasVierling: Bool, highPairRank: Rank?, secondPairRank: Rank?, highTripleRank: Rank?) {
+
+    var hasPair: Bool = false
+    var hasSecondPair: Bool = false
+    var hasTriple: Bool = false
+    var hasVierling: Bool = false
+    var highPairRank: Rank? = nil
+    var highTripleRank: Rank? = nil
+    var secondPairRank: Rank? = nil
+    
+    for i in 1...4 {
+        if cards[i].rank == cards[i-1].rank {
+            if !hasPair {
+                // finde erstes Pärchen
+                hasPair = true
+                highPairRank = cards[i].rank
+            }
+            else { // 2tes paar oder drilling oder vierling
+                hasSecondPair = true
+                secondPairRank = cards[i].rank
+            }
+            if (i < 4 && cards[i + 1].rank == cards[i].rank) { //Drilling gefunden
+                hasTriple = true
+                highTripleRank = cards[i].rank
+            }
+            if(i < 3 && cards[i + 1].rank == cards[i].rank && cards[i + 2].rank == cards[i].rank) { // Vierling gefunden
+                hasVierling = true
+            }
+        }
+    
+    }
+    return (hasPair, hasSecondPair, hasTriple, hasVierling, highPairRank, secondPairRank, highTripleRank)
+}
+
+
+func HasStraight (cards : [Card]) -> Bool {
+    var hasStraight: Bool = false
+    var i: Int = 1
+    while i < 5 {
+        if 1 + cards[i].rank.rawValue ==  cards[i-1].rank.rawValue {
+            i += 1
+            
+        } else {
+            break
+            
+        }
+        if i == 4 {hasStraight = true}
+    }
+    return hasStraight
+}
+
```

Comment:
Sehr gut, alle `Ranking`s implementiert 👏 und übersichtlich in Funktionen ausgelagert 👍

Statt der globalen Funktionen könntest du die Tests lieber als Computed Attributes in `PokerHand` schreiben. Außerdem Funktionen klein schreiben, nur Datentypen groß ;)